src/board.rs

```diff
@@ -16,12 +16,68 @@ use ratatui::{
 };
 use uci::Engine;
 
+// only the pure gameboard, no additional information
+pub type GameBoard = [[Option<(PieceType, PieceColor)>; 8]; 8];
+
+#[derive(PartialEq, Clone, Debug, Eq, PartialOrd, Ord, Copy)]
+pub struct Coord {
+    /// row, line, y
+    pub row: u8,
+    /// column, x
+    pub col: u8,
+}
+impl Coord {
+    pub fn new<T: Into<u8>>(row: T, col: T) -> Self {
+        Coord {
+            row: row.into(),
+            col: col.into(),
+        }
+    }
+    /// optional new: try making a valid [`Coord`], if can't, return [`None`]
+    pub fn opt_new<T: TryInto<u8>>(row: T, col: T) -> Option<Self> {
+        let row: u8 = row.try_into().ok()?;
+        let col: u8 = col.try_into().ok()?;
+
+        let ret = Coord { row, col };
+        if !ret.is_valid() {
+            None
+        } else {
+            Some(ret)
+        }
+    }
+    /// not yet set position, has to later be set and only used afterwards
+    pub fn undefined() -> Self {
+        Coord {
+            row: UNDEFINED_POSITION,
+            col: UNDEFINED_POSITION,
+        }
+    }
+    /// checks whether `self` is valid as a chess board coordinate
+    pub fn is_valid(&self) -> bool {
+        (0..8).contains(&self.col) && (0..8).contains(&self.row)
+    }
+}
+
+impl std::ops::Index<&Coord> for GameBoard {
+    type Output = Option<(PieceType, PieceColor)>;
```

Comment:
This can be the new GameBoard type

Reply:
It wont let me add a separate comment since it isn't edited code but same applies for `pub fn is_piece_opposite_king(piece: Option<(PieceType, PieceColor)>, color: PieceColor) -> bool {` in utils.rs on line 229.

Reply:
I'm sorry, but I'm afraid, you're mistaken.
The `GameBoard` looks like: `[[Option<(PieceType, PieceColor)>; 8]; 8]`, while the ones you suggested to change are only `Option<(PieceType, PieceColor)>` which we don't (yet) have a seperate `type` for.
Please correct me if I'm wrong.